llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/ContainerRunnerImpl.java

```diff
@@ -342,6 +352,33 @@ public SubmitWorkResponseProto submitWork(SubmitWorkRequestProto request) throws
         .build();
   }
 
+  // if request is coming from llap external client, verify the JWT
+  // as of now, JWT contains applicationId
```

Comment:
which applicationId is it? I assume llap application ID? can you explain a little bit more about how/who injects the llap application ID into jwt and how llap validate it against its own app ID.

Reply:
added a comment in code which explains this - 

In GenericUDTFGetSplits

// 6. Generate JWT for external clients if it's a cloud deployment
        // we inject extClientAppId in JWT which is same as what fragment contains.
        // extClientAppId in JWT and in fragment are compared on LLAP when a fragment is submitted.
        // see method ContainerRunnerImpl#verifyJwtForExternalClient

    
In ContainerRunnerImpl#verifyJwtForExternalClient


// extClientAppId is injected in JWT and fragment request by initial get_splits() call.
      // so both of these - extClientAppIdFromJwt and extClientAppIdFromSplit should be equal eventually if the signed JWT is valid for this request.
      // In get_splits, this extClientAppId is obtained via LlapCoordinator#createExtClientAppId which generates a
      // application Id to be used by external clients.